File: src/test/java/me/nasukhov/TukitaLearner/bot/command/AnswerQuestionTest.java
package me.nasukhov.TukitaLearner.bot.command;

import me.nasukhov.TukitaLearner.bot.io.*;
import me.nasukhov.TukitaLearner.study.*;
import org.junit.jupiter.api.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.transaction.annotation.Transactional;

import java.util.*;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

@SpringBootTest
@Transactional
class AnswerQuestionTest {
    @Autowired AnswerQuestion handler;
    @Autowired AskQuestion askQuestion;
    @Autowired ProgressTracker progressTracker;
    @Autowired GroupQuestionRepository groupQuestionRepository;

    @Autowired
    private QuestionRepository questionRepository;

    private Output output;

    @BeforeEach
    void setup() {
        output = mock(Output.class);

        questionRepository.save(
            new Question("2+5 equals to", "seven", new ArrayList<>(){{
                add("thirteen");
                add("seven");
                add("twenty-one");
            }})
        );
    }

    @AfterEach
    void reset() {
        questionRepository.deleteAll();
    }

    @Test
    void handleCorrect() {
        var channel = new Channel("SomeChannelId");
        var user = new User("SomeId", "SomeName");

        var questionResult = progressTracker.createRandomForChannel(channel);
        assertTrue(questionResult.isPresent());
        var newGroupQuestion = questionResult.get();
        assertFalse(newGroupQuestion.hasAnswerFromUser(user.name()));

        var questionId = newGroupQuestion.getId().toString();
        var correctAnswerId = getCorrectVariant(newGroupQuestion);

        Input input = new Input(String.format("qh answer %s %d", questionId, correctAnswerId), channel, user);

        handler.handle(input, output);

        verify(output).write(String.format("%s, да, правильный ответ «<spoiler>%s</spoiler>».", user.name(), newGroupQuestion.viewAnswer()));
    }

    private int getCorrectVariant(GroupQuestion groupQuestion) {
        var variantId = 0;

        var correctAnswer = groupQuestion.viewAnswer();
        while (!groupQuestion.viewVariant(variantId).equals(correctAnswer)) {
            if (variantId > 1000) {
                throw new RuntimeException("Looks like an endless loop");
            }

            variantId++;
        }

        return variantId;
    }
}

File: src/test/kotlin/me/nasukhov/TukitaLearner/bot/command/AnswerQuestionTest.kt
package me.nasukhov.TukitaLearner.bot.command

import me.nasukhov.TukitaLearner.bot.io.Channel
import me.nasukhov.TukitaLearner.bot.io.Input
import me.nasukhov.TukitaLearner.bot.io.Output
import me.nasukhov.TukitaLearner.bot.io.User
import me.nasukhov.TukitaLearner.study.*
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.transaction.annotation.Transactional

@SpringBootTest
@Transactional
internal class AnswerQuestionTest(
    @Autowired private val handler: AnswerQuestion,

    @Autowired private val progressTracker: ProgressTracker,

    @Autowired private val questionRepository: QuestionRepository
) {
    private lateinit var output: Output

    @BeforeEach
    fun setup() {
        output = Mockito.mock(Output::class.java)

        questionRepository.save(
            Question(
                "2+5 equals to",
                "seven",
                listOf("thirteen", "seven", "twenty-one")
            )
        )
    }

    @AfterEach
    fun reset() {
        questionRepository.deleteAll()
    }

    @Test
    fun handleCorrect() {
        val channel = Channel("SomeChannelId")
        val user = User("SomeId", "SomeName")

        val questionResult = progressTracker.createRandomForChannel(channel)
        Assertions.assertTrue(questionResult.isPresent)
        val newGroupQuestion = questionResult.get()
        Assertions.assertFalse(newGroupQuestion.hasAnswerFromUser(user.name))

        val questionId: String? = newGroupQuestion.id.toString()
        val correctAnswerId = getCorrectVariant(newGroupQuestion)

        val input = Input(String.format("qh answer %s %d", questionId, correctAnswerId), channel, user)

        handler.handle(input, output)

        Mockito
            .verify(output)
            .write("${user.name}, да, правильный ответ «<spoiler>${newGroupQuestion.viewAnswer()}</spoiler>».")

        handler.handle(input, output)

        Mockito.verifyNoMoreInteractions(output)
        // TODO check that answer is recorded exactly once
    }

    private fun getCorrectVariant(groupQuestion: GroupQuestion): Int {
        var variantId = 0

        val correctAnswer = groupQuestion.viewAnswer()
        while (groupQuestion.viewVariant(variantId) != correctAnswer) {
            if (variantId > 1000) {
                throw RuntimeException("Looks like an endless loop")
            }

            variantId++
        }

        return variantId
    }
}
